storage/src/mmr/mutable.rs

```diff
@@ -0,0 +1,512 @@
+//! An in-memory authenticatable & updatable key-value store.
+//!
+//! A _mutable MMR_ is an authenticatable and updatable key-value store based on an [Mmr] over all
+//! updates called the _update tree_, and an authenticatable [Bitmap] that indicates which leaves in
+//! the update tree correspond to the latest value of each key. For example, if the same key K was
+//! updated at positions X, Y, and Z in that order, then the bitmap will store 0s for the bits
+//! corresponding to X and Y, and a 1 for the bit corresponding to Z. The root hash of the mutable
+//! MMR is the result of hashing together the roots of these two structures.
+
+use crate::mmr::{
+    bitmap::Bitmap, hasher::Hasher, iterator::leaf_pos_to_num, mem::Mmr,
+    verification::KeyValueProof, Error,
+};
+use commonware_cryptography::Hasher as CHasher;
+use commonware_utils::Array;
+use std::{collections::HashMap, marker::PhantomData};
+use tracing::error;
+
+/// A record representing one key/value update or deletion.
+pub struct UpdateRecord<K: Array, V: Array> {
+    /// The position of the leaf representing this record in the updates tree, or Bitmap::DELETED if
+    /// this record corresponds to a deletion of the key.
+    tree_pos: u64,
+
+    /// The key that was updated or deleted.
+    key: K,
+
+    /// The value the key was assigned in this update. For deletions, the value is the previously
+    /// active value for the key.
+    value: V,
+}
+
+/// A mutable MMR based key-value store.
+pub struct MutableMmr<K: Array, V: Array, H: CHasher> {
+    /// An MMR over digests of the key/value pairs from each UpdateRecord. The number of leaves in
+    /// this MMR always equals the number of records.
+    updates: Mmr<H>,
+
+    /// An authenticatable bitmap where a 1 indicates the corresponding update record contains the
+    /// active value for its key. The bit count of this bitmap always equals the number of records.
+    is_active: Bitmap<H>,
+
+    /// All updates and deletions made to the store in order of execution. The index of each record
+    /// in this vector is called its _location_.
+    records: Vec<UpdateRecord<K, V>>,
+
+    /// A map from each key to the record location containing its most recent update. Keys in the
+    /// deleted state are not stored in this map.
+    snapshot: HashMap<K, usize>,
+}
+
+impl<K: Array, V: Array, H: CHasher> MutableMmr<K, V, H> {
+    /// DELETED is used as a deletion indicator. Its value is guaranteed to be reserved for use as
+    /// either a record location or a tree position.
+    const DELETED: u64 = 0;
+
+    /// Create a new, empty mutable MMR. The `key` and `value` are used for initialization -- their
+    /// specific values are unimportant other than behaving as seeds for the root hash.
+    pub fn new(hasher: &mut H, key: K, value: V) -> Self {
```

Comment:
I suppose this is a byproduct of `Array` not having a default value? Feels a bit weird to require this for new but I suppose it could be the way you "namespace" your root?

Alternative approach could be to make `Record` an enum and have an option with no K/V (just position) and one with just K (for deletion). This could also enable us to put "markers" at the edges of a batch (if that became useful).

Reply:
Correct, it's really due to lack of defaulting. This is the structure we will eventually store on disk, so didn't want to make it too fancy with Rust-specific types like enums thinking it might complicate things?

Reply:
As discussed, it is doable (given each enum a byte type) but doesn't help with using fixed-size entries.